### DISCUSSION TITLE
is remix not for large projects ?

### DISCUSSION THREAD

**Discussion by shop-fluencer** [remix-run/remix#5044] [Q&A]
Hello,

I have read an article and have read this:

Hello,

I have read an article and have read this:

> Remix version 1 is excellent for internal projects and small company websites. Full-stack engineers should take the opportunity to take a deep dive into the documentation, sample projects, and Remix source code. There is an incredible amount of information to learn from.

So is remix not recommend for medium/large projects now ?

article: https://moduscreate.com/blog/remix-what-you-should-know-from-our-experience/

**Comment by dmarkow** [comment]
Maybe they're just seeing the "v1" and assuming it's not ready, but we're using it in a large project approaching 1,000 routes. It's *way* easier to manage our app using Remix's structure, especially the loaders and actions, than it ever was with the mess of our webpack, manually managed routes/layouts, waterfall data loading with apollo, a custom-but-fragile form handling approach, difficult SSR, etc... the list goes on. I can just focus on shipping features now.

To top it all off, our app is (or at least feels) much faster since I was able to remove about 95% of the waterfall data and loading spinners.

**Reply by tombohub** [reply]
are you comparing to client side app? Can you compare with frameworks like django , laravel and rails?

**Comment by kiliman** [comment]
Hmm... they make that claim "_Remix version 1 is excellent for internal projects and small company websites_", but they don't state any reasons for that opinion.

There is nothing inherently limiting with Remix that you can't use it to build any React-based app.

**Comment by sergiodxa** [comment]
I have also been using it on a large fintech project for almost a year with a better DX (and UX) than I had previously with other frameworks, just because someone wrote it's for internals projects and small companies doesn't mean it's for that.

Also take into account that Shopify acquired it to replace Hydrogen core with Remix and probably use it for Shopify itself, and Shopify is one of the larges tech companies.

**Reply by tombohub** [reply]
which other frameworks? How do you compare it with django, laravel and rails?

**Comment by nickdirienzo** [comment]
Hope it's okay that I'm replying to a month+ old post 😇 

@dmarkow @sergiodxa awesome to hear that y'all are both using it for large projects! Have either of y'all written about how you structure your app to support many routes (in @dmarkow's case over 1,000 of them!)? 

I definitely find it delightful to build features in Remix and excited for our app to grow! I'm new to Remix and am curious to learn how folks are organizing large-scale projects.

**Reply by dmarkow** [reply]
Honestly, we're not doing anything all that different from smaller apps.

- We have nested routes based on the business purpose, and no single route folder/parent has more than 4-5 routes directly in it. So our "/projects/..." parent has around 130 routes alone, most of them under "/projects/:number", which almost acts as its own app from there (technically if we wanted to, we could break this into separate apps, but I really don't see the benefit).
- The main place you see that it's a large app is in the top level routes folder -- we have around 60 top level routes there -- but inside each folder it's super manageable.
- All shared components sit in `app/components` and queries (we're still on GraphQL for our pre-existing API) are in `app/queries`, but I'll move us to the new flat routes by folder structure soon, so components and queries used just by a single route can be colocated with that route.
- We have around 250 resource routes that a lot of separate parts of our app use -- data fetching and saving from various modals and dropdowns, mostly. Those are all super simple routes with just either a loader or action.
- The only noticeable impact of our project being so large is dev reload times. A standard save/rebuild/reload loop takes around 3 seconds for our app right now. It's a bit slower than our old webpack/react app (I seem to recall 500-600ms reloads on our app there), but I imagine more performance improvements will hit Remix as time goes on (I've already submitted a couple PRs for things specific to large apps). And like I mentioned in my original reply, the actual end-user experience of the app is much better than before.

**Reply by nickdirienzo** [reply]
Thanks so much for the fast reply @dmarkow! 🙏 

Re colocating by domain / business purpose: that sounds like y'all have implemented where my intuition went. Awesome to hear and glad it's manageable at the folder level. I was a little worried about `routes` directory getting too hard to explore but it sounds doable at 60 top-level routes.

Thanks for the additional insight re component and query management! That's what we have too, at least on the component side of things. 

I'm excited to also try out flat routes to further colocate non-route modules. 

I appreciate you taking the time to share your perspective here. Thanks again!

**Reply by zsrobinson** [reply]
In regards to colocating non-route modules, I've personally just been specifying that all files that start with an underscore are excluded from being routed inside of `remix.config.js`. No flat routes required!

It looks something like this:

```js
module.exports = {
  ignoredRouteFiles: ["**/.*", "**/_*"],
};
```

Not sure where I originally saw this, but it's been great not having to unnecessarily move things into `app/components` when I don't have to. If using loader data, it's still just as easy to import the loader from the `index.tsx` file in the same directory to pass to the `useLoaderData` hook (if I'm not mistaken, you just import it from `"."`).

**Reply by sergiodxa** [reply]
@nickdirienzo I gave a talk at Remix Conf EU about how I use Remix at my work, you can [find it on YouTube](https://www.youtube.com/watch?v=eh6FfHFpYMM&list=PLEDCVpGG3y_PQpwWHwPKXhLvM5ZfsgGQA).

**Reply by nickdirienzo** [reply]
@zsrobinson great tip! Thank you. I do like the idea of putting not system components and other non-shared functionality closer to the route logic. 

@sergiodxa thanks for sharing! Great talk; that was really helpful. And also your blog has been super informative + helpful -- thanks for writing down what you're learning and building. I recently implemented the Policy authorization functions and it really simplifies things.

**Reply by ChayaninSuatap** [reply]
@dmarkow Thank you for sharing. It's very nice to see a serious Remix adoption.

I can see myself that with more route files, the rebuild time get slower. For me I've just written 13 routes and the reload time took about 4 second already.

I am so afraid that with more source code added to the project, the rebuild time will grow bigger.
I have experimented creating like a thousand routes file in project and the rebuild time took 15 second. It seems rebuild time grow with number of files.

Do you have any particular config to improve rebuild time?

**Reply by dmarkow** [reply]
@ChayaninSuatap So rebuilds definitely slow down as the project size increases. I did get a PR merged recently that will help rebuild times quite a bit (it drops mine from 12-13 seconds down to 3-4), it didn't make it into 1.16.0 but it should be out in the next release.

**Comment by barbinbrad** [comment]
I would even go so far to say that bigger projects will get more benefit out of remix. With most apps, performance starts to degrade as the project gets larger. But with remix I’m getting the same 100 Lighthouse performance score with hundreds of routes and lots of nesting that I was getting when I had a few routes.

**Reply by ClimenteA** [reply]
After npm run build what are the steps for serving in prod?

npm run build then npm run start - are not enough?

I get ~60 performance on lighthouse and some react minify errors in console.

**Reply by barbinbrad** [reply]
This is a platform-dependent question. I'd recommend starting with one of the stacks. I use vercel, for example, which just requires that you run the build command.

**Reply by tombohub** [reply]
why is performance degrading with large apps that's not affecting remix?
